File: problem2-bfs.py
from collections import deque
class TrieNode:
    def __init__(self):
        self.word = ""
        self.child = [False for _ in range(26)]


class Solution:
    
    def __init__(self):
            self.root = TrieNode()
            
    def insert(self, word):

        curr = self.root
        for i in range(len(word)):
            idx = ord(word[i]) - ord("a")
            if not curr.child[idx]:
                curr.child[idx] = TrieNode()
            curr = curr.child[idx]
        curr.word = word 
        
        
    def longestWord(self, words: List[str]) -> str:
        
        for word in words:
            self.insert(word)
        
        q = deque()
        
        q.append(self.root)
        
        while q:
            size = len(q)
            for i in range(size):
                curr = q.popleft()
                for j in range(25, -1, -1):
                    if curr.child[j] and curr.child[j].word:
                        q.append(curr.child[j])
                        
                        
        return curr.word
                        
#Time complexity and Space complexity O(NL) N -> size of input L -> average length of word                        
                        
                        
                        
                
        
            
                

File: Problem2_BFS.java
//Time Complexity: O(n*l) ; where n is no. of final length words & l is their length.
//Space Complexity: O(n*l)
//Code run successfully on LeetCode.

public class Problem2_BFS {

    class TrieNode{
        
        TrieNode[] children;
        String word;
        
        TrieNode(){
            
            children = new TrieNode[26];
        }
    }
    
    TrieNode root = new TrieNode();
    
    private void insert(String word){
        
        TrieNode curr = root;
        
        for(int i =0; i < word.length(); i++){
            
            char c = word.charAt(i);
            
            if(curr.children[c - 'a'] == null)
                curr.children[c - 'a'] = new TrieNode();
            
            curr = curr.children[c - 'a'];
        }
         curr.word = word;  
    }
    
    public String longestWord(String[] words) {
        
        for(String word : words)
            insert(word);
        
        Queue<TrieNode> q = new LinkedList<>();
        q.add(root);
        TrieNode curr = root;
        
        while(!q.isEmpty()){
            
            curr = q.poll();
            
            for(int i = 25; i >=0 ; i--){
                
                if(curr.children[i] != null && curr.children[i].word != null)
                    q.add(curr.children[i]);
                
            }
        }
        
        if(curr.word == null)
            return "";
        
        return curr.word;
    }
}
